src/box/memtx_tx.c

```diff
@@ -2247,152 +2276,164 @@ memtx_tx_history_prepare_insert_stmt(struct txn_stmt *stmt)
+		 * complicated and deserves an explanation with example.
+		 * Imagine a space with primary index (pk) by the first field
+		 * and secondary index (sk) by the second field.
+		 * Imagine then three in-progress transactions that executes
+		 * replaces {1, 1, 1}, {2, 1, 2} and {1, 1, 3} correspondingly.
+		 * What must happen when the first transaction commits?
+		 * Both other transactions intersect the current in the sk.
+		 * But the second transaction with {2, 1, 2} must be aborted
+		 * (or sent to read view) because of conflict: it now introduces
+		 * duplicate insertion to the sk.
+		 * On the other hand the third transactions with {1, 1, 3} has
+		 * a right to live since it tends to overwrite {1, 1, 1} in
+		 * both pk and sk.
+		 * To handle those conflicts in general we must scan chains
+		 * towards the top and check insert statements.
+		 */
+		struct memtx_story *newer_story = story;
+		while (newer_story->link[i].newer_story != NULL) {
+			newer_story = newer_story->link[i].newer_story;
+			struct txn_stmt *test_stmt = newer_story->add_stmt;
+			/* Don't conflict own changes. */
 			if (test_stmt->txn == stmt->txn)
 				continue;
+			/*
+			 * Ignore case when other TX executes insert after
+			 * precedence delete.
+			 */
 			if (test_stmt->is_own_change &&
 			    test_stmt->del_story == NULL)
 				continue;
-			if (test_stmt->del_story == story)
-				continue;
-			memtx_tx_handle_conflict(stmt->txn, test_stmt->txn);
 			/*
-			 * Note that it's a secondary index and there's no
-			 * need to call memtx_tx_story_link_deleted_by since
-			 * all is done by the primary story chain.
+			 * Ignore the case when other TX overwrites in both
+			 * primary and secondary index.
 			 */
-		}
-		/* Now link is in top of chain, where gap records are stored. */
-		struct gap_item *item, *tmp;
-		rlist_foreach_entry_safe(item, &link->read_gaps,
-					 in_read_gaps, tmp) {
-			if (item->txn == stmt->txn || item->is_nearby)
-				continue;
-			memtx_tx_handle_conflict(stmt->txn, item->txn);
-		}
-	}
-
-	for (uint32_t i = 0; i < index_count; i++) {
-		old_story = story->link[i].older_story;
-		if (old_story == NULL)
-			continue;
-
-		struct tx_read_tracker *tracker;
-		rlist_foreach_entry(tracker, &old_story->reader_list,
-				    in_reader_list) {
-			if (tracker->reader == stmt->txn)
+			if (test_stmt->del_story == story)
 				continue;
-			memtx_tx_handle_conflict(stmt->txn, tracker->reader);
+			memtx_tx_handle_conflict(stmt->txn, test_stmt->txn);
 		}
+		/*
+		 * We have already checked gap readers before for the case
+		 * of insertion to the primary index.
+		 * In any (replace or insert) case we must handle gap readers
+		 * in the secondary indexes as well since in all kinds of
+		 * statements can insert new value to secondary index.
+		 * Note that newer_story is in top of chain due to previous
+		 * manipulations.
+		 */
+		memtx_tx_handle_conflict_gap_readers(newer_story, i, stmt->txn);
 	}
```

Comment:
It would be a bit inconsistent: all other calls to `memtx_tx_handle_conflict` in `memtx_tx_history_prepare_insert_stmt` will be inlined, except for this one. So I decided to leave it simple while in special further commit "memtx: simplify and refactor memtx_tx_history_prepare_stmt" I moved all conflict handling to two helper functions: `memtx_tx_handle_conflict_story_readers` and `memtx_tx_handle_conflict_gap_readers`. Please see that commit and ensure that it looks very natural there.